Plane: send attitude target message implementation (for quadplanes)

miodine opened the pull request:
Minor mod to the communication routine, whose purpose is to expose the target attitude for a plane operating in VTOL mode.

Might be useful for external in-flight PID tuning applications.

IamPete1 commented on ArduPlane/GCS_Mavlink.cpp:
```diff
@@ -157,6 +157,48 @@ void GCS_MAVLINK_Plane::send_attitude() const
+#if HAL_QUADPLANE_ENABLED
+    plane.quadplane.attitude_control->get_throttle_in();
+
+    if (!(plane.g2.flight_options & FlightOptions::GCS_REMOVE_TRIM_PITCH_CD)) {
+        target_p -= radians(plane.g.pitch_trim_cd * 0.01f);
+    }
```
This does not do anything.

miodine replied:
right, deleted

IamPete1 commented on ArduPlane/GCS_Mavlink.cpp:
```diff
@@ -157,6 +157,48 @@ void GCS_MAVLINK_Plane::send_attitude() const
+    Quaternion quat(0.0f,0.0f,0.0f,0.0f);
+
+#if HAL_QUADPLANE_ENABLED
+    plane.quadplane.attitude_control->get_throttle_in();
```
```suggestion
    throttle = plane.quadplane.attitude_control->get_throttle_in();
```

IamPete1 reviewed (dismissed):
This seems to have collected some extra unrelated changes. It only need to be sent if quadplane attitude control is active, 


You should be able to more or less copy exactly the copter implementation.

IamPete1 reviewed (commented):
You can use `last_att_control_ms` to check if its valid to report, like we do for PID logging:

https://github.com/ArduPilot/ardupilot/blob/fcd4152b30aa1f50b110303e7cae1f9f35f5e645/ArduPlane/Log.cpp#L30

So if not active you can early return

Then you can just copy and paste the copter implementation to get the values, just swapping `copter.` for `plane.quadplane.`

https://github.com/ArduPilot/ardupilot/blob/fcd4152b30aa1f50b110303e7cae1f9f35f5e645/ArduCopter/GCS_Mavlink.cpp#L98-L100

IamPete1 commented on ArduPlane/GCS_Mavlink.cpp:
```diff
@@ -159,33 +159,25 @@ void GCS_MAVLINK_Plane::send_attitude() const
-    float throttle = 0.0f;
-    Quaternion quat(0.0f,0.0f,0.0f,0.0f);
+    // Check if the attitude target is valid for reporting
+    if (AP_HAL::millis() - plane.quadplane.last_att_control_ms > 100) return;
```
You can stash millis in a const and use it down on 175 to save getting it twice. Return should be on the next line per the style guide.
```suggestion
    const uint32_t now = AP_HAL::millis()
    if (now  - plane.quadplane.last_att_control_ms > 100) {
        return;
    }
```

miodine replied:
Right, of course, I completely missed the other function call

IamPete1 commented on ArduPlane/GCS_Mavlink.cpp:
```diff
@@ -159,33 +159,25 @@ void GCS_MAVLINK_Plane::send_attitude() const
 
     mavlink_msg_attitude_target_send(
         chan,
         AP_HAL::millis(),       // time since boot (ms)
```
```suggestion
        now,       // time since boot (ms)
```

IamPete1 reviewed (commented):
Code changes now look good, just need to drop the logging changes and squash the commits into one with a sensible name. `Plane: send attitude target for quadplanes` or similar.

IamPete1 commented on ArduPlane/GCS_Mavlink.cpp:
```diff
@@ -157,6 +157,35 @@ void GCS_MAVLINK_Plane::send_attitude() const
 }
 
+void GCS_MAVLINK_Plane::send_attitude_target() {
+
```
Sorry, I just remembered something. We need to define this out if quadplane is not enabled.
```suggestion
#if HAL_QUADPLANE_ENABLED
```

and a `#endif // HAL_QUADPLANE_ENABLED` at the end

IamPete1 commented on ArduPlane/GCS_Mavlink.cpp:
```diff
@@ -157,6 +157,35 @@ void GCS_MAVLINK_Plane::send_attitude() const
+void GCS_MAVLINK_Plane::send_attitude_target() {
+
+    // Check if the attitude target is valid for reporting
+    const uint32_t now = AP_HAL::millis()
```
```suggestion
    const uint32_t now = AP_HAL::millis();
```

miodine commented:
I did as it was requested - I applied the most recent suggestions, reverted any chages done in the logging part, and squashed the commits altogether.

IamPete1 reviewed (approved):
I think this looks good. Its not added to any default stream rate, so it will not be sent by default, but it can be requested.

IamPete1 commented:
@miodine I have marked for the dev cal which is this evening on Discord, your welcome to come along. 

https://ardupilot.org/dev/docs/ardupilot-discord-server.html#weekly-dev-calls

miodine commented:
All right, then I'll await the results of your discussion. Sorry for my clumsiness with all that, and thanks for the guidance

tridge commented on ArduPlane/GCS_Mavlink.cpp:
```diff
@@ -157,6 +157,39 @@ void GCS_MAVLINK_Plane::send_attitude() const
+        return;
+    }
+    
+    const Quaternion quat  = plane.quadplane.attitude_control->get_attitude_target_quat();
```
should be a const ref in AC_AttitudeControl.h, and ref here

tridge commented on ArduPlane/GCS_Mavlink.cpp:
```diff
@@ -157,6 +157,39 @@ void GCS_MAVLINK_Plane::send_attitude() const
+    }
+    
+    const Quaternion quat  = plane.quadplane.attitude_control->get_attitude_target_quat();
+    const Vector3f ang_vel = plane.quadplane.attitude_control->get_attitude_target_ang_vel();
```
should be const ref

tridge reviewed (changes requested):
fix refs in copter too?

tridge commented:
this needs to be against master, not plane-4.4

tridge commented:
@IamPete1 will rework this as a PR against master
